### PR TITLE
Support OpenTelemetry @WithSpan in CDI

### PR DESCRIPTION
**Pull request by radcortez**
A few considerations:
 - The OpenTelemetry instrumentation BOM includes a lot of integration stuff with other libraries we use. We didn't try these and may cause some confusion to someone including them in their project via our BOM.
 - We probably require some more test, but wanted to push just to start the discussion on the implementation.

### CODE REVIEW

**Review by ebullient — APPROVED** [review]

**Review by kenfinnigan — COMMENTED** [review]
Are there benefits of using the Instrumentor libraries vs interacting with the tracer directly?

File: extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/TracerProducer.java [hunk 1]
```diff
@@ -25,6 +27,6 @@ public LateBoundSampler getLateBoundSampler() {
     @Singleton
     @DefaultBean
     public Tracer getTracer() {
-        return GlobalOpenTelemetry.getTracer("io.quarkus.opentelemetry");
+        return GlobalOpenTelemetry.getTracer(INSTRUMENTATION_NAME);
```
[the remark this answers is not in the document]
Comment (radcortez) [reply extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/TracerProducer.java:30]: Well, that seems to be one of the ideas from the discussion here in the OTel side when named Tracers were introduced: https://github.com/open-telemetry/opentelemetry-specification/issues/10.

It also allows you to identify which library created the span. I think it is fine.

**Comment by radcortez** [comment]
> Are there benefits of using the Instrumentor libraries vs interacting with the tracer directly?

The benefit here is that we only need to configure the `Instrumenter` and define where to start and stop the `Span` and the `Instrumenter` will take care of setting the tags, picking and propagating other tags etc. They can also be enabled or disabled. When I was looking into the OTel SDK this seems to be the proper way to implement manual instrumentation: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/using-instrumenter-api.md

I also wrote a prototype in SR OpenTelemetry to try it out first: https://github.com/smallrye/smallrye-opentelemetry/tree/main/implementation

My plan was to refactor our manual `Tracer` code in `Vert.x` and use the `Instrumenter` instead.

**Comment by kenfinnigan** [comment]
My understanding was manual instrumentation was done with https://github.com/open-telemetry/opentelemetry-java, the OTeL Java SDK directly, and not through the `Instrumentor` classes in the agent repository.

**Comment by radcortez** [comment]
I believe that the Instrumeter APIs were not mature enough in the past. They seem to be better now and the way to go. For instance if you look into their Spring Web implementation they do it use the Instrumenter API: https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/spring/spring-web-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/web

**Comment by ebullient** [comment]
Right.. but that is instrumentation agent... not something like quarkus (that would run outside/above/alongside the agent). Granted, we're kind of halfway in-between being an agent and not..

**Comment by kenfinnigan** [comment]
I would recommend not to use auto instrumentation APIs in non-agent usage in Quarkus, but it's up to you @radcortez

**Comment by radcortez** [comment]
> Right.. but that is instrumentation agent... not something like quarkus (that would run outside/above/alongside the agent). Granted, we're kind of halfway in-between being an agent and not.

My understanding is that the Instrumenter API is intended to be used as an encapsulator to all the logic of gathering Telemetry data (either agent or not). The added benefit is that you can easily register the appropriate extractors and do not have to manually reimplement the specification.

If you take the example I gave from the Spring implementation it is implemented using a REST interceptor: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/spring/spring-web-3.1/library/src/main/java/io/opentelemetry/instrumentation/spring/web/RestTemplateInterceptor.java.

There are other modules where the API is used as a simple library. Actually it follows a pattern where if a module is named `library` inside the OTel implementation it is named this way. Here are a few other examples:

- JDBC Driver: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/ed02aff7b5aab9475ed590024ced9e9bde3f13e1/instrumentation/jdbc/library/src/main/java/io/opentelemetry/instrumentation/jdbc/datasource/OpenTelemetryDataSource.java#L40
- CXF: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/90c0df9328b87aa8fb7a4c928dedae1d18ea15c1/instrumentation/jaxws/jaxws-2.0-cxf-3.0/library/src/main/java/io/opentelemetry/javaagent/instrumentation/cxf/CxfHelper.java#L18
- GRPC: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/9de4041d4c703aa5d9600c92b8c4f53cb5d9dff3/instrumentation/grpc-1.6/library/src/main/java/io/opentelemetry/instrumentation/grpc/v1_6/TracingServerInterceptor.java#L24

If we are not sure we can certainly clarify in the OTel call if the API is intended to be used this way.

**Comment by radcortez** [comment]
> I think my question is whether or not they are ready for the instrumentation APIs to be used this way outside of that repo (i.e. do they think it is stable).

That is something we can confirm / clarify in one of the OTel calls.

> Then we (quarkus) have another question to answer differently, which is whether or not any of those libraries could/should work as-is.

I believe we should follow the same rule as any other library. If it is not a Quarkus supported extension, then you are on your own. Agents are a different story.

**Comment by radcortez** [comment]
In the last OTel call, I've confirmed with @kenfinnigan that the Instrumenter API is also intended to be used for libraries.

@kenfinnigan are you in favour to proceed with this approach?

**Comment by kenfinnigan** [comment]
Entirely up to you now @radcortez!

Only comment I'd make is whether it also requires revisiting the existing implementation to utilize the same API for consistency.

**Comment by radcortez** [comment]
> Only comment I'd make is whether it also requires revisiting the existing implementation to utilize the same API for consistency.

Yes, I plan to do it as well for Vert.x in another PR. Thanks!

**Pull request merged**